### ISSUE TITLE
Why don't we use MSE as a reconstruction loss for VAE ?

### ISSUE THREAD

**Issue by ahmed-fau** [pytorch/examples#399]
Hi,

I am wondering if there is a theoretical reason for using BCE as a reconstruction loss for variation auto-encoders ? Can't we simply use MSE or norm-based reconstruction loss instead ?

Best Regards

**Comment by nwesemann** [comment]
+1

**Comment by kingofspace0wzz** [comment]
According to the original VAE paper[1], BCE is used because the decoder is implemented by MLP+Sigmoid which can be viewed as a 'Bernoulli distribution'. You can use MSE if you implement a Gaussian decoder. Take the following pseudocode for an example, 
```
mu = MLP(z)
sigma = MLP(z)
reconstruction = Gaussian(mu, sigma)
```


Ref

[1] https://arxiv.org/abs/1312.6114

**Comment by ahmed-fau** [comment]
> because the decoder is implemented by MLP+Sigmoid which can be viewed as a 'Bernoulli distribution'.

Does this mean that in order to model "continuous" distribution like Gaussian then we should not use sigmoid as the output layer and  replace it with tanh or even flattened conv layer for example ?

**Comment by ahmed-fau** [comment]
Then using conv layer and flatten it could be a possible solution for modelling output layer ?

**Comment by muammar** [comment]
> > because the decoder is implemented by MLP+Sigmoid which can be viewed as a 'Bernoulli distribution'.
> 
> Does this mean that in order to model "continuous" distribution like Gaussian then we should not use sigmoid as the output layer and replace it with tanh or even flattened conv layer for example ?

If I change https://github.com/pytorch/examples/blob/master/vae/main.py#L60 from ` .sigmoid()` to `.tanh()`, and https://github.com/pytorch/examples/blob/master/vae/main.py#L74 from `BCE` to `MSE`, will that make this VAE to try a Gaussian Reconstruction and work for [-1, 1]? I would appreciate any input.

**Comment by rasoulMrz** [comment]
> > > because the decoder is implemented by MLP+Sigmoid which can be viewed as a 'Bernoulli distribution'.
> > 
> > 
> > Does this mean that in order to model "continuous" distribution like Gaussian then we should not use sigmoid as the output layer and replace it with tanh or even flattened conv layer for example ?
> 
> If I change https://github.com/pytorch/examples/blob/master/vae/main.py#L60 from ` .sigmoid()` to `.tanh()`, and https://github.com/pytorch/examples/blob/master/vae/main.py#L74 from `BCE` to `MSE`, will that make this VAE to try a Gaussian Reconstruction and work for [-1, 1]? I would appreciate any input.

Using the MSE loss to model a Gaussian distribution needs the output to be unbounded. writing the loss equation will help you understand this better.
in order to model a Gaussian distribution, you will use the output of your model as an estimator for the mean. therefore you don't have to add a tanh (no activations needed). keep in mind that by using a Gaussian distribution you are not using the fact that your data is bounded between -1 and 1 so there is always a probability for the model to generate values outside this range.
If your data is in the range [-1, 1], you can also scale and shift the data to [0, 1] and keep using the Sigmoid+BCE.

**Comment by GPla** [comment]
@muammar To approximate a gaussian posterior, it usually works fine to use no activation function in the last layer and interpret the output as mean for a normal distribution. If we assume a constant variance for the posterior, we naturally end up with the MSE as loss function. An alternative option is proposed by [An et al.](https://www.semanticscholar.org/paper/Variational-Autoencoder-based-Anomaly-Detection-An-Cho/061146b1d7938d7a8dae70e3531a00fceb3c78e8). We can duplicate the output layer of the decoder to model the mean and variance of the normal distribution and then optimize the negative log likelihood.
